### PR TITLE
[SPARK-17667] [YARN][WIP]Make locking fine grained in YarnAllocator#enqueueGetLossReasonRequest

### PR DESCRIPTION
**Pull request by ashwinshankar77**
## What changes were proposed in this pull request?

Following up on the discussion in SPARK-15725, one of the reason for AM hanging with dynamic allocation(DA) is the way locking is done in YarnAllocator. We noticed that when executors go down during the shrink phase of DA, AM gets locked up. On taking thread dump, we see threads trying to get loss for reason via YarnAllocator#enqueueGetLossReasonRequest, and they are all BLOCKED waiting for lock acquired by allocate call. This gets worse when the number of executors go down are in the thousands, and I've seen AM hang in the order of minutes. This patch is created to make the locking little more fine grained by remembering the executors that were killed via AM, and then serve the GetExecutorLossReason requests with that information.

This patch is not final. I want to get inputs on how I can go about removing executors from the `executorsKilledByDriver` set. Also if there is a better way to solve this, would be happy to make the changes.
## How was this patch tested?

This was tested in our cluster with manually scaling the number of executors to thousands and then shrinking them down to something small, and made sure that we don't see BLOCKED threads stuck at YarnAllocator#enqueueGetLossReasonRequest.

### CODE REVIEW

**Comment by vanzin** [comment]
ok to test

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 1]
```diff
@@ -522,7 +525,9 @@ private[yarn] class YarnAllocator(
             // This is usually happened when explicitly killing a container, the result will be
             // returned in one AM-RM communication. So query RPC will be later than this completed
             // container process.
-            releasedExecutorLossReasons.put(eid, exitReason)
+            if (!executorsKilledByDriver.contains(eid)) {
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: there is a race here where if the executor went down for reason other then us kill after we called kill we won't get the right loss reason.  ie killExecutor just adding it to list to be removed later, if it happened to die for another reason we would miss out on that.   I don't think this is a big deal but we could put in a check here to compare reason.

**Comment by tgravescs** [comment]
SPARK-17365 should help with this situation have you tried that?

overall the approach seems ok to me.  Its an easy short cut to prevent asking if we already know. There is the possible race condition but it doesn't seem very big issue since it should have been idle and if it happens to die for some other reason after we requested to kill it we probably don't care.

**Comment by tgravescs** [comment]
It looks like you should be able to remove it after the context.reply(ExecutorKilled) return.  The killExecutors should be called before the loss reason and if there is a race there its just going to go to yarn to get the reason, which goes back to the same issue but I assume that will be rare.  We shouldn't be asking for the loss reason more then once.

**Comment by ashwinshankar77** [comment]
Thanks for the review @tgravescs! I'll update the patch shortly. We are also trying out SPARK-17365.

**Comment by vanzin** [comment]
@ashwinshankar77 do you plan on updating this and removing the "WIP"?

**Comment by HyukjinKwon** [comment]
Hi @ashwinshankar77, if you are not currently able to work on this further, maybe it should be closed for now. It seems inactive for few months.

**Comment by deshanxiao** [comment]
I meet this problem in spark 2.1 just now. RPC threads are all blocked by the lock.

**Pull request closed without merging**